Join order benchmark (IMDB) validation and benchmark improvements

Tmonster opened the pull request:
This PR has 2 separate improvements, and basically re-opens 

1. Create a benchmark to run and verify IMDB. We weren't doing this before, the results were gathered using a script I set up here (https://github.com/Tmonster/verify_db_scripts). I plan on adding more to this repository soon. In order to properly verify with postgres, I had to modify the IMDB queries to return all lines (and not a min), with an order so that the COLLATE and the languages wouldn't produce differrent results.

2. Allow benchmarks to create and connect to a cached duckdb instance so that cached data doesn't need to be exported to parquet and imported for every query for every benchmark run. Results on improvements can be found in the table below. Timing information excludes the cost to download benchmark data, as that is very network dependent. 

A benchmark template can call connect {db_path}. The benchmark runner will then attempt to connect to the db at {db_path}. If an error is thrown, this is most likely because of storage format instability. When this is the case, the db is removed, and a connection is made to a clean instance. If no exception is thrown, the database is checked for tables. If no tables exist, data will be loaded in from duckdb_benchmark_data/{{cache}}.

If tables do exist, the benchmark will run as normal. If the data is incorrect, the benchmarks will report incorrect results fail. At this point it is up to the user to fix the data or answers, most likely by deleting the cached database.

Why do you need the full path from from the duckdb directory?
- If someone wants to connect to a duckdb that isn't in the duckdb_benchmark_data/{{cache}} folder, they can easily do that. 


| Benchmark  | Cached DB size | largest cached parquet file (uncompressed to CSV) | with "connect" | without "connect" |
| ------------- | ------------- | ------------- | ------------- | ------------- |
| TPC-H (sf=1) | 262MB  | 719MB  | 9.507s  | 20.465s  |
| TPC-DS (sf=1)| 313MB  | 381MB  | 1:06 min  | 2:00 min  |
| clickbench  | 2.8GB  | 7.3G  | q19-q22 hang  | OOM locally  |
| h2oai-join  | 1.1G  | 440MB  | 19s  | 20.335s  |
| h2oai-group  | 584MB  | 466MB  | 17.139  | 18.806  |
| JOB  | 2GB | 915MB  | 3:11 min  | 7:05 min  |


TPC-H, TPC-DS, and IMDB all have improvements while h2oai regresses. I have noticed some weird behavior for clickbench where the size of the database grows between q19-q22 even though the queries are only select queries. Currently it hangs on my work computer, but I'll investigate next week on a more powerful computer.

 I could amend the code to open the connection in read only mode to prevent changes, but then it requires more logic when the db turns out to not have data and the benchmark runner needs to then delete the database and re-connect to load data.

Commit:
Issue #5277: Sorted Aggregate Sorting

Disable busted test.

Tishj committed:
successfully treating union as a struct for now

Tishj committed:
changed to actually using a Union type and properly displayiong the contents of the UNION vector

Mytherin committed:
Merge pull request #5290 from LindsayWray/enum_type_not_by_name

Enum type added to the types metadata table

Mytherin committed:
Merge pull request #5185 from krlmlr/f-styler

R: code format

Maxxen committed:
refactored control flow

Tishj committed:
removed union from test_all_types

Tishj committed:
actually removed

Tishj committed:
Merge branch 'master' into julia_union_type

papparapa committed:
fix to use memcmp in starts_with function

Tishj committed:
Merge branch 'feature' into julia_union_type

Mytherin committed:
Merge branch 'master' into feature

Tishj committed:
Merge branch 'feature' into julia_union_type

Tmonster committed:
force CI to run again

Tishj committed:
fix hidden merge issues

Tishj committed:
Merge branch 'master' into truncating_decimal

Tishj committed:
Merge branch 'feature' into truncating_decimal

Tishj committed:
fix bug with negatives, added tests for it

Mytherin committed:
Lateral joins MVP

Mytherin committed:
Lateral joins with explicit LATERAL syntax working

Mytherin committed:
LATERAL Left joins working

Mytherin committed:
More lateral join tests

Mytherin committed:
way more lateral join test cases

Mytherin committed:
Create filter out of arbitrary expressions for lateral join

Mytherin committed:
More lateral join tests with aggregates

Mytherin committed:
Throw error message on nested lateral joins (not yet supported)

Mytherin committed:
Format

Tishj committed:
add explicit type for the enum class, move the enum class and the RoundInfo struct out of DecimalCastData

Mytherin committed:
Lateral cross join working now as well

Mytherin committed:
Fixes for binding lateral joins

Commit:
Merge branch 'orderagg-sorter' of github.com:hawkfish/duckdb into orderagg-sorter

Commit:
Issue #5277: Sorted Aggregate Sorting

Replace ChunkCollection with ColumnDataCollection.
Buffer small argument counts to avoid CDC allocation for small groups.

Tmonster committed:
adding proper imdb benchmark

Tmonster committed:
add benchmark CMakeLists

Tmonster committed:
forcing another CI run

Mytherin committed:
Merge pull request #5334 from papparapa/add-starts-with-function-and-operator

Add starts_with function and operator

Mytherin committed:
Fix for non-lateral joins/cross products in correlated subqueries: the lateral binder introduces an additional level of depth to expressions (as there is one extra expression binder in the stack) - as such we now need to decrease the depth of those expressions again

Mytherin committed:
Handle recursive subqueries in ExpressionDepthReducer

Mytherin committed:
Fixes for single-file compilation

Mytherin committed:
Amalgamation fix

Tmonster committed:
added more imdb benchmark files

Tmonster committed:
fixed some answers

Tishj committed:
remove Value::SetNull function

Tishj committed:
remove nested struct from DecimalCastData because that turns it into a 'non-trivial' type

Mytherin committed:
Add more lateral join tests, and avoid duplicate-elimination when dealing with lists (to pre-emptively prevent the same issue with lateral joins as we ran into with subqueries, #3593)

Mytherin committed:
Add missing tsv

Mytherin committed:
Add missing tenk.tsv file

Tmonster committed:
fix benchmark scripts. Still need to test incorrect versions for cached dbs

Mytherin committed:
Non-temp table for force_reload

Commit:
Merge branch 'timezone-function' of github.com:hawkfish/duckdb into timezone-function

Mytherin committed:
Merge pull request #5337 from Maxxen/feature/string-binary-cast

Feature: Allow binary-formatted strings to be cast to integers

Tmonster committed:
fix the load script

Mytherin committed:
Merge branch 'master' into feature

Commit:
Issue #5314: At Time Zone

Add coverage tests.

Commit:
Merge branch 'master' into orderagg-sorter

Tishj committed:
remove constructor, to stop DecimalCastData from being a non-trivial type

Commit:
Issue #5277: Sorted Aggregate Sorting

Fix DataChunk lazy initialisation.

Mytherin committed:
Merge pull request #5399 from Mytherin/nljoinisnotalwaysterrible

For range joins use NL join when the LHS or RHS side is tiny

Tmonster committed:
added another test requested by mark

Mytherin committed:
Merge pull request #5393 from Mytherin/lateraljoins

Add support for LATERAL joins

Mytherin committed:
Merge pull request #5360 from Tishj/julia_union_type

[Julia] Add support for consuming a UNION vector into a DataFrame

Commit:
read sorted result in parallel

Tmonster committed:
tpcds benchmark template update as well

Tmonster committed:
add imdb to regressions

Mytherin committed:
Merge pull request #5341 from hawkfish/timezone-function

Issue #5314: At Time Zone

Mytherin committed:
Merge branch 'master' into feature

Commit:
add test for order parallel CTAS with persistent storage

hannes committed:
hey this does something

Tishj committed:
update psqlodbc checkout hash

Tmonster committed:
force CI again

hannes committed:
no uuids

hannes committed:
added check that prevents writing to existing directories

Commit:
add union test

lokax committed:
Supoort enum_code() function

Commit:
Merge branch 'master' into orderagg-sorter

Commit:
Merge branch 'orderagg-sorter' of https://github.com/hawkfish/duckdb into orderagg-sorter

Commit:
Issue #5277: Sorted Aggregate Slicing

Slice in place.

hannes committed:
compile fix an csv test

hannes committed:
forbidding combination of copy flags

lokax committed:
Add prepare stmt test

Commit:
remove optimization from physical batch insert (messes up order preservation)

Commit:
Merge branch 'master' into parallel_orderby_getdata

Commit:
undo changes (create separate PR)

Mytherin committed:
Shell: add individual SQL queries to the history, instead of individual lines

Mytherin committed:
Add support for history search to shell

Mytherin committed:
Handling of tab and Ctrl+G

Mytherin committed:
History search: duplicate eliminate entries

Mytherin committed:
Simplify rendering of search match

Commit:
Issue #5277: Sorted Aggregate Scans

Use Scans correctly (with Reset()).

Mytherin committed:
Windows fix

Mytherin committed:
Remove the "search" text from the search rendering

Mytherin committed:
Merge pull request #5362 from Tishj/truncating_decimal

Decimal values now round when the value given has more decimals than the `scale` of the target

Mytherin committed:
Merge pull request #5414 from Mytherin/multilinehistory

Shell: add individual SQL queries to the history, instead of individual lines

Mytherin committed:
Merge pull request #5415 from Mytherin/shellhistorysearch

Shell: add support for history search

satotake committed:
httpfs: check environment vars for AWS Credentials

- Support AWS environment variables
    - AWS_DEFAULT_REGION
    - AWS_ACCESS_KEY_ID
    - AWS_SECRET_ACCESS_KEY
    - AWS_SESSION_TOKEN
- issue #4021

hannes committed:
hopefully fixing ci

Tmonster committed:
again force CI. There's no reason for the old init to go wrong, seeing as h2oai also makes an https call to github. Maybe the plan_cost_runner doesnt compile duckdb with httpfs?

Tmonster committed:
fix format errors

Tmonster committed:
more format fix changes I think'

Tishj committed:
Merge branch 'feature' into duckframes

Mytherin committed:
Merge pull request #5403 from lnkuiper/parallel_orderby_getdata

Parallelise scanning result of ORDER_BY

Mytherin committed:
Merge pull request #5212 from zhouliqi/master

Add translate function

changhiskhan committed:
Enable cmake to recognize AppleClang

If the compiler id is AppleClang, many of the cmake branches that
check for strequal Clang fails (e.g., static linking no longer
supported). This changes the build to just check whether the
string ends in compiler id ends in Clang.

Tishj committed:
Merge branch 'feature' into duckframes

Tmonster committed:
fix becnhmark runner so storage works on CI

Tmonster committed:
h2oai is faster when benchmarks are in memory

Tishj committed:
add 'description' to Relation

Mytherin committed:
Merge branch 'master' of github.com:duckdb/duckdb into feature

Mytherin committed:
Merge branch 'feature' of github.com:duckdb/duckdb into feature

Mytherin committed:
Merge branch 'master' into feature

Mytherin committed:
Merge branch 'feature' of github.com:duckdb/duckdb into feature

Tishj committed:
undo changes to merge result->relation

Tishj committed:
remove leftovers of removed changes

hannes committed:
fine no csv tests

Tmonster committed:
benchmarks are looking a lot better. Let us see what CI/CD does

Tmonster committed:
fix some other issues

Tishj committed:
added missing include

Tmonster committed:
again a CI fix/check

Tmonster committed:
change some benchmark files to use a cached tpch db

Tmonster committed:
format fix changes

Tishj committed:
add missing include

Mytherin committed:
Merge pull request #5432 from changhiskhan/master

Enable cmake to recognize AppleClang

Mytherin committed:
Merge pull request #5408 from lokax/enum-code

Support enum_code() function

Mytherin committed:
Merge pull request #5302 from Tmonster/another_binder_error_2

Fix binder error and produce more informative error message.

Commit:
Merge branch 'master' into orderagg-sorter

Commit:
Merge branch 'orderagg-sorter' of github.com:hawkfish/duckdb into orderagg-sorter

Commit:
Issue #5277: Sorted Aggregate Benchmark

Add a new benchmark for large numbers of groups.
Performance seems unchanged from the old sort code.

Commit:
Issue #5258: Quantile Negative Fractions

Extend the fractions for quantiles to accept negative values in [-1, 0].
This new syntax uses DESC sorting to compute the half-open intervals.
It also make PERCENTILE_XXX DESC compatible with Oracle and PG.

Mytherin committed:
Re-use Parquet buffers in column readers

Mytherin committed:
Compilation fixes

Mytherin committed:
More descriptive allocation message

Mytherin committed:
DUCKDB_API

Tishj committed:
missing header for 'std::reverse'

Mytherin committed:
Merge branch 'feature' of github.com:duckdb/duckdb into feature

Mytherin committed:
Merge branch 'master' into feature

Mytherin committed:
Add support for reading RLE-encoded boolean columns to the Parquet reader

hannes committed:
first stab

ila committed:
print profiling output for deserialized logical query plans

hannes committed:
taking arrowarrystream directly to get out of leaking nightmare

Mytherin committed:
DELTA_BYTE_ARRAY encoding working

Mytherin committed:
Support DELTA_LENGTH_BYTE_ARRAY

Mytherin committed:
Add spark v2 tests

ila committed:
format fix

Mytherin committed:
DELTA_BYTE_ARRAY/DELTA_LENGTH_BYTE_ARRAY tests

Tishj committed:
fix up some warnings (treated as errors)

Mytherin committed:
Fix for skipping junk repetition_bytes if set in V2

Tishj committed:
fix a tidy-check 'issue'

Tishj committed:
address a couple of tidy-check issues, regarding push_back in loop

Mytherin committed:
Merge pull request #5445 from Mytherin/parquetreusebuffer

Parquet Reader: Re-use (de)compression and dictionary buffers and allocate powers of two

Tishj committed:
replaced ugly code with glob

Commit:
Merge branch 'master' into orderagg-sorter

Commit:
Merge branch 'orderagg-sorter' of https://github.com/hawkfish/duckdb into orderagg-sorter

Commit:
Merge branch 'master' into quantile-desc

Commit:
Merge branch 'master' into quantile-disc

Commit:
Merge branch 'quantile-disc' of github.com:hawkfish/duckdb into quantile-disc

Commit:
Merge branch 'quantile-disc' into quantile-desc

Mytherin committed:
Merge branch 'feature' of github.com:cwida/duckdb into feature

Commit:
Issue #5366: Negative Quantile Sorting

Add parentheses to try to fix strange Linux failures.

Commit:
Issue #5366: Negative Quantile Sorting

Add parentheses to try to fix strange Linux failures.

Commit:
Issue #5366: Quantile Absolute Value

Use our AbsOperator instead of whatever Linux is using today.
Make ICU TZ sorting deterministic.

Commit:
Issue #5366: Quantile Absolute Value

Make ICU TZ sorting deterministic.

Commit:
Issue #5366: QUANTILE_DESC Decimals

Maintain decimal types to avoid floating point scaling errors.

Commit:
Issue #5366: QUANTILE_DESC Templates

Fix the templating so Linux will swallow it.

hannes committed:
don't need that destructor

Mytherin committed:
Merge pull request #5457 from Mytherin/deltabytearray

Support RLE, DELTA_BYTE_ARRAY and DELTA_LENGTH_BYTE_ARRAY Parquet encodings

Mytherin committed:
Correctly set internal flag for internal function catalog entries, and expose the internal flag in duckdb_functions

Mytherin committed:
Add "allow_internal" flag to alter entry, and remove duplicate code path for adding new overloads to an existing scalar function

Tishj committed:
resolve includes for pythonpkg

Mytherin committed:
Merge pull request #5448 from ila/master

print profiling output for deserialized logical query plans

Mytherin committed:
Merge pull request #5456 from hawkfish/orderagg-sorter

Issue #5277: Sorted Aggregate Sorting

Tishj committed:
clang-tidy now sees the pythonpkg files, and the includes are set up correctly

Tishj committed:
add the same clang-tidy boilerplate to all pythonpkg subfolder cmakelists

Tishj committed:
(hopefully) fixed clang-tidy for pythonpkg

Commit:
Merge branch 'master' into quantile-desc

Tishj committed:
fixed tidy-check issues

Commit:
Issue #5366: QUANTILE_DESC Simplification

Use a copy constructor instead of double negation.
(As Bjarne intended...)

Tishj committed:
re-run CI? - code-cov failed to upload

Tishj committed:
moved FromValue out of PyResult

Tishj committed:
not reserving for enough

hannes committed:
this is the correct way

Mytherin committed:
Merge pull request #5462 from Mytherin/functioninternal

Add internal flag to duckdb_functions, and correctly set internal flag for internal functions

Mytherin committed:
Merge branch 'master' into feature

hannes committed:
no catalog addition

Tmonster committed:
initial commit

Tmonster committed:
some code cleanup

Tishj committed:
Merge branch 'feature' into duckframes

Tishj committed:
Merge branch 'master' into duckframes

hannes committed:
merge

hannes committed:
documentation

hannes committed:
Merge pull request #5479 from hannes/rstringsusertype

Add experimental R String passthrough support

papparapa committed:
Add python from_parquet features

Mytherin committed:
Merge pull request #5463 from hawkfish/quantile-desc

Issue #5258: Quantile Negative Fractions

Tishj committed:
Merge branch 'master' into duckframes

Tmonster committed:
force CI again. IMDB shouldn't run out of data

Mytherin committed:
Merge branch 'master' into feature

Tmonster committed:
some code cleanup and better error expectations

samansmink committed:
added retry to refresh possible failing keep-alive conn

samansmink committed:
format

Tmonster commented:
IMDB regression fails because the old benchmark runner doesn't set a path for temporary storage, and the worker doesn't have enough memory to store the whole imdb database

satotake committed:
use `SetOption`

papparapa committed:
Collapse four overloads into two

papparapa committed:
Merge branch 'master' of https://github.com/duckdb/duckdb into add-python-from-parquet-features

papparapa committed:
Fix __init__.pyi

hannes committed:
Merge pull request #5449 from hannes/jdbcarrowingest

Arrow stream ingestion for JDBC client

hannes committed:
Merge pull request #5412 from hannes/per_thread_output

PER_THREAD_OUTPUT flag for COPY

Mytherin committed:
Merge branch 'master' into feature

Tishj committed:
Merge branch 'feature' into duckframes

papparapa committed:
Merge branch 'duckdb:master' into add-python-from-parquet-features

Commit:
add extract statements capi

samansmink committed:
rm unused var

Mytherin committed:
Skip broken tests for now

samansmink committed:
unified http retry mechanism for all basic requests

ListObjectV2 requests are not retried as these are never long running

samansmink committed:
cleanup, change http retry defaults

Mytherin committed:
Merge pull request #5532 from Mytherin/disabletests

Feature: skip broken tests for now

Mytherin committed:
Merge branch 'master' into feature

samansmink committed:
format

Mytherin committed:
Merge branch 'feature' of github.com:duckdb/duckdb into feature

Commit:
Merge branch 'feature' into extract_statements_capi

Commit:
feature branch merge

hannes committed:
Merge pull request #5484 from Tmonster/union_all_relation_R_3

Add Union All support to R extention

Tmonster commented:
IMDB regression fails because the old duckdb benchmark build can't load the imdb database into memory. When this merges, imdb regression tests should pass

Mytherin commented on benchmark/include/interpreted_benchmark.hpp:
```diff
@@ -15,6 +15,8 @@ namespace duckdb {
 struct BenchmarkFileReader;
 class MaterializedQueryResult;
 
+static const string DEFAULT_DB_PATH = "duckdb_benchmark_db.db";
```
Is this used anywhere?

Mytherin commented on benchmark/interpreted_benchmark.cpp:
```diff
@@ -395,12 +412,13 @@ void InterpretedBenchmark::Cleanup(BenchmarkState *state_p) {
 
 string InterpretedBenchmark::GetDatabasePath() {
-	if (!InMemory()) {
+	if (!InMemory() && cache_db.compare("") == 0) {
```
nit: `cache_db.empty()`

Mytherin commented on benchmark/interpreted_benchmark.cpp:
```diff
@@ -395,12 +412,13 @@ void InterpretedBenchmark::Cleanup(BenchmarkState *state_p) {
 		return path;
-	} else if (db_path != "") {
-		return db_path;
+	} else if (cache_db.compare("") != 0) {
```
Nit: `cache_db.empty()`

Mytherin reviewed (commented):
Thanks for the PR! Looks great. Some minor comments below.

> I have noticed some weird behavior for clickbench where the size of the database grows between q19-q22 even though the queries are only select queries.

Is that still happening? It might have something to do with the temporary table creation. I noticed earlier that creating temporary tables can trigger checkpoints in persistent databases (as the transaction mistakenly thinks changes have been made to the database).

Tmonster committed:
ok, addressed last set of PR comments

Commit:
add out of bounds test

Tmonster commented:
@Mytherin From what I could tell the DB expansion doesn't happen anymore. The temp table issue fixed it

samansmink committed:
now counting correctly

Tishj committed:
merge branch 'feature' into duckframes

Mytherin committed:
Merge pull request #5492 from papparapa/add-python-from-parquet-features

[Python] Add from_parquet features

samansmink committed:
Apply suggestions from code review

typos

Co-authored-by: Elliana May <me@mause.me>

samansmink committed:
nicer error messages on connection errors

Commit:
Issue #5277: Sorted Aggregate Window

Add a throwing Window function to the SortedAggregateFunction wrapper.
This will prevent crashes if we ever accidentally try to window it.

Mytherin committed:
Merge branch 'master' into feature

Mytherin committed:
Format fix

Mytherin committed:
Merge branch 'feature' of github.com:duckdb/duckdb into feature

Commit:
Issue #5559: 2022g Time Zones

Update to 2022g ICU time zone data drop.

Tishj committed:
formatting

Mytherin committed:
Merge pull request #5524 from LindsayWray/extract_statements_capi

Add ExtractStatements to C API

Mytherin committed:
Merge pull request #5549 from samansmink/improve-http-retry

Improve http retry

Commit:
Remove blank line that breaks format check.

Commit:
Merge branch 'feature' into timezone-2022g

Commit:
Merge branch 'feature' into ordaggr-window

Commit:
Merge branch 'feature' into quantile-decimal

Commit:
Merge branch 'feature' of github.com:duckdb/duckdb into feature

Commit:
Merge branch 'feature' into timezone-2022g

Commit:
Merge branch 'feature' into quantile-decimal

Commit:
Merge branch 'feature' into ordaggr-window

Mytherin committed:
Merge pull request #5571 from hawkfish/ordaggr-window

Issue #5277: Sorted Aggregate Window

Mytherin committed:
Merge pull request #5572 from hawkfish/quantile-decimal

Issue #5422: QUANTILE_DESC Decimals

Commit:
Merge branch 'feature' into timezone-2022g

Mytherin committed:
Merge pull request #5570 from hawkfish/timezone-2022g

Issue #5559: 2022g Time Zones

Mytherin committed:
Merge pull request #5436 from Tishj/duckframes

[Dev] Clean up of the python pkg folder structure

Mytherin committed:
Merge pull request #5419 from satotake/httpfs-set-aws-envvar

httpfs: check environment vars for AWS Credentials

Tmonster committed:
force CI

Tmonster committed:
delete default db path
